src/lib/autoinstall/ask/runner.rb

```diff
@@ -0,0 +1,215 @@
+    #
+    # ## How does it work?
+    #
+    # The process starts by grouping all the questions defined in the <ask-list> section
+    # by their `dialog` ID. All the questions with the same ID will be placed in the same
+    # dialog. If an ID is not specified, the question is shown on a separate dialog.
+    #
+    # Then the first dialog is presented to the user. When she/he presses the 'Next'
+    # button, this runner processes the answers and goes to the next dialog. Processing
+    # the response implies:
+    #
+    # * Updating the profile if an AutoYaST path was given (elements `path` and
+    #   `pathlist` from the profile).
+    # * Writing the answer to the file, if a `file` element was specified.
+    # * Running the given script (`script` element from the profile).
+    #
+    # @example Ask questions from the profile during the 1st stage
+    #   Yast::Profile.ReadXML("autoinst.xml")
+    #   runner = Runner.new(profile, stage: :initial)
+    #   runner.run
+    class Runner
+      # @return [Hash] AutoYaST profile
+      attr_reader :profile
+      # @return [Symbol] Stage to run
+      attr_reader :stage
+
+      # @param profile [Hash] AutoYaST profile
+      # @param stage [Symbol] :initial or :cont
+      def initialize(profile, stage: :initial)
+        @profile = profile
+        @indexes = []
+        @stage = stage
+      end
+
+      # Runs the dialogs and processes the user input
+      #
+      def run
+        current_dialog = go_next
+        loop do
+          break if current_dialog.nil?
+
+          result = run_dialog(current_dialog)
+          if result == :back
+            current_dialog = go_back
+          elsif result == :next
+            current_dialog.questions.each { |q| process_question(q) }
+            current_dialog = go_next
+          end
+        end
+      end
+
+    private
+
+      # Returns the AskListSection from the profile
+      #
+      # @return [AskListSection]j
+      def ask_list
+        return @ask_list if @ask_list
+
+        section = profile.fetch("general", {}).fetch("ask-list", [])
+        @ask_list = Y2Autoinstall::AutoinstProfile::AskListSection.new_from_hashes(section)
+      end
+
+      NEXT_DIALOG_FILE = "/tmp/next_dialog".freeze
+
+      # Run the given dialog
+      #
+      # @param dialog [Y2Autoinstall::Ask::Dialog] Ask dialog specification
+      # @return [Symbol] :next or :back
+      def run_dialog(dialog)
+        ask_dialog = Y2Autoinstall::Widgets::AskDialog.new(
+          dialog, disable_back_button: first?
+        )
+        ask_dialog.run
+      end
+
+      # Finds the ID of the next dialog
+      #
+      # It checks for a file located at {NEXT_DIALOG_FILE} which contains the number
+      # of the following dialog. If it does not exist, it goes with the dialog after
+      # the current one.
+      #
+      # @return [Integer,nil] Next dialog ID
+      def find_next_dialog_index
+        return unless ::File.exist?(NEXT_DIALOG_FILE)
```

Comment:
I expect this won't be liked by security team as it is fix file location which you read and rm so especially in second stage there is some window to attack. ( not sure how serious it is as there is probably not much users in given time frame ). Maybe we should do at least trivial satify checks like it is only regular file (to prevent malicious symlinks) and its size is reasonable (to prevent memory kill).

Reply:
of course as location is documented and known to users, we cannot change it.